### PR TITLE
using binary search in findBreakpoint() and insertBreakpoint()

### PR DESCRIPTION
**Pull request by pvmm**
binary searching breakpoints is much more efficient, since they are already ordered by address value.

### CODE REVIEW

**Review by m9710797 — CHANGES_REQUESTED** [review]
I think there's a bug in findBreakpoint():
lower_bound() may return breakpoints.end(), but on line 323 you unconditionally dereference 'i'.

File: src/DebuggerData.cpp [hunk 1]
```diff
@@ -271,13 +271,12 @@ bool Breakpoints::inCurrentSlot(const Breakpoint& bp)
...
 {
-	for (auto it = breakpoints.begin(); it != breakpoints.end(); ++it) {
-		if (it->address > bp.address) {
-			breakpoints.insert(it, bp);
-			return;
-		}
-	}
-	breakpoints.push_back(bp);
+	auto it = std::upper_bound(breakpoints.begin(), breakpoints.end(), bp,
+	           [](const Breakpoint& bp1, Breakpoint bp2) {
```
Comment (m9710797) [inline src/DebuggerData.cpp:275]: Detail: also pass 'bp2' by const-reference.

**Comment by m9710797** [comment]
Thanks for this PR! I'll probably apply it (after the above review comments have been resolved).

Nitpick: I don't agree with your description "binary search is much more efficient". For few breakpoints (less than 10?) linear search is probably faster. And I think typically there won't be many breakpoints. But of course binary search scales much better to larger collections. So I think it's a good idea.

File: src/DebuggerData.cpp (hunk 1 above)
Comment (pvmm) [reply src/DebuggerData.cpp]: fixed

**Comment by pvmm** [comment]
> I think there's a bug in findBreakpoint(): lower_bound() may return breakpoints.end(), but on line 323 you unconditionally dereference 'i'.

I tested the code and it worked as expected, returning -1 when not found. But now I've put an extra check.

**Comment by pvmm** [comment]
> Thanks for this PR! I'll probably apply it (after the above review comments have been resolved).
> 
> Nitpick: I don't agree with your description "binary search is much more efficient". For few breakpoints (less than 10?) linear search is probably faster. And I think typically there won't be many breakpoints. But of course binary search scales much better to larger collections. So I think it's a good idea.

I think for any "serious use" of the debugger it is welcome.

**Comment by m9710797** [comment]
> I tested the code and it worked as expected, returning -1 when not found. But now I've put an extra check.
Indeed most of the time it will work as expected: dereferencing the end-iterator will most likely give a value different than 'addr', and then, most likely, this function will return -1. But you can get unlucky and get the wrong result. Or the iterator could point past the end of the allocated memory page and then you get a segmentation fault.

Anyway, it's fixed now. Thanks. I'll merge your PR.

**Pull request merged**